### DISCUSSION TITLE
Fetch token via authorization response URL at later point in time

### DISCUSSION THREAD

**Discussion by autochargetesla** [tdorssers/TeslaPy#71] [General]
Hello!

Excellent work on this library !

As described in the section [Alternative](https://github.com/tdorssers/TeslaPy#alternative), the URL generated after successful Tesla login in the browser is (immediately) supplied to `authorization_response`. As a result, the tokens are generated and available to TeslaPy. This works perfectly !

In our use case, there will be a minor delay between these two steps. Therefore, the question is: **is it possible to run TeslaPy to generate the tokens from the `authorization_response` URL independently from the first step?** 

To summarize this two-step use case:

1. Run TeslaPy to generate `tesla.authorization_url()`, paste URL in the browser and perform Tesla login resulting in 404 page. Copy this URL for later use.
2. Run TeslaPy (again) and supply the URL copied before as `authorization_response` (or elsewhere?) to generate set the of tokens (e.g. refresh token).

Cheers!

**Comment by tdorssers** [comment]
This is currently not possible. I have to change the return type of authorization_url() from a string to a tuple that contains the url and state string, as this is required for the two-step use case to work.

**Reply by autochargetesla** [reply]
Thank you for the response. With url and state string you mean `authorization_url` and `state` correct? Simply restoring those values (as obtained as a result of step 1) by setting them in the `tesla` object does not seem to work (causes `Unauthorized client`). 

It also requires `code_verifier` to be set. Then it works.

**Comment by tdorssers** [comment]
This should work:

```python
import teslapy

tesla = teslapy.Tesla('elon@tesla.com')
if not tesla.authorized:
    print('Use browser to login. Page Not Found will be shown at success.')
    state = tesla.new_state()
    url = tesla.authorization_url(state=state)
    code_verifier = tesla.code_verifier
    print('Open this URL: ' + url)
tesla.close()
del tesla

tesla = teslapy.Tesla('elon@tesla.com', state=state)
if not tesla.authorized:
    tesla.fetch_token(authorization_response=input('Enter URL after authentication: '),
                      code_verifier=code_verifier)
vehicles = tesla.vehicle_list()
print(vehicles[0])
tesla.close()
```

Does this cover your use case?

**Reply by autochargetesla** [reply]
Hi! 

Thank you very much :-) 

This is exactly what we needed!

Have a wonderful Easter !

Cheers

**Reply by tdorssers** [reply]
Commit e88ec08 includes better support for staged authorization and an updated readme as well.
Happy Easter

**Comment by autochargetesla** [comment]
Hello @tdorssers, thank you for your previous support on answering my questions!

I have a follow-up question:
Are you aware of any expiration date in the future of `authorization_url` or will it always be valid?

Cheers

**Reply by tdorssers** [reply]
It should always be valid
